window/src/os/x11/modifiers.rs

```diff
@@ -0,0 +1,345 @@
+        log::debug!("Key: {:?}", key);
+
+        if algo.failed {
+            return;
+        }
+
+        if algo.try_shift {
+            if key == algo.shift_keycode {
+                return;
+            }
+
+            state.update_key(algo.shift_keycode, xkb::KeyDirection::Down);
+        }
+
+        let changed_type = state.update_key(key, xkb::KeyDirection::Down);
+
+        if changed_type
+            & (xkb::STATE_MODS_DEPRESSED | xkb::STATE_MODS_LATCHED | xkb::STATE_MODS_LOCKED)
+            != 0
+        {
+            let mods = state.serialize_mods(if algo.try_shift {
+                xkb::STATE_MODS_EFFECTIVE
+            } else {
+                xkb::STATE_MODS_DEPRESSED | xkb::STATE_MODS_LATCHED | xkb::STATE_MODS_LOCKED
+            });
+
+            let keysyms = state.key_get_syms(key);
+            log::debug!("Keysym: {:x?}", keysyms);
+
+            macro_rules! S2 {
+                ($key_l:expr, $key_r:expr, $mod:ident) => {
+                    if keysyms[0] == $key_l || keysyms[0] == $key_r {
+                        if algo.$mod == 0 {
+                            log::debug!(
+                                "Keycode {:?} triggered keysym '{:x?}' for {:?}' modifier.",
+                                key,
+                                keysyms[0],
+                                stringify!($mod)
+                            );
+                            algo.$mod = mods;
+                        } else if algo.$mod != mods {
+                            log::debug!(
+                                "Keycode {:?} triggered again keysym '{:x?}' for {:?}' modifier.",
+                                key,
+                                keysyms[0],
+                                stringify!($mod)
+                            );
+                            algo.failed = true;
+                        }
+                    }
+                };
+            }
+
+            macro_rules! S1 {
+                ($key:expr, $mod:ident) => {
+                    if keysyms[0] == $key && algo.$mod == 0 {
+                        log::debug!(
+                            "Key {:?} triggered keysym '{:?}' for '{:?}' modifier.",
+                            key,
+                            keysyms[0],
+                            stringify!($mod)
+                        );
+                        algo.$mod = mods;
+                    }
+                };
+            }
+
+            // We can handle exactly one keysym with exactly one bit set in the implementation
+            // below; with a lot more gymnastics, we could set up an 8x8 linear system and solve
+            // for each modifier in case there are some modifiers that are only present in
+            // combination with others, but it is not worth the effort.
+            if keysyms.len() == 1 && mods.is_power_of_two() {
+                S2!(xkb::keysyms::KEY_Shift_L, xkb::KEY_Shift_R, shift);
```

Comment:
why does just this one have `xkb::keysyms` as the namespace and the others are `xkb::`?